obspy/core/stream.py

```diff
@@ -3118,6 +3119,80 @@ def remove_sensitivity(self, *args, **kwargs):
             tr.remove_sensitivity(*args, **kwargs)
         return self
 
+    def stack(self, group='all', type='linear', npts_tol=0):
+        """
+        Return stream with traces stacked by the same selected metadata.
+
+        :param str group: Stack waveforms together which have the same metadata
+            given by this parameter. The parameter should name the
+            corresponding keys of the stats object,
+            e.g. ``'{network}.{station}'`` for stacking all
+            locations and channels of the stations and returning a stream
+            consisting of one stacked trace for each station.
+            This parameter can take two special values,
+            ``'id'`` which stacks the waveforms by SEED id and
+            ``'all'`` (default) which stacks together all traces in the stream.
+        :type type: str or tuple
+        :param type: Type of stack, one of the following:
+            ``'linear'``: average stack (default),
+            ``('pw', order)``: phase weighted stack of given order,
+            see [Schimmel1997]_,
+            ``('root', order)``: root stack of given order.
+        :param int npts_tol: Tolerate traces with different number of points
+            with a difference up to this value. Surplus samples are discarded.
+
+        :returns: New stream object with stacked traces. The metadata of each
+            trace (inlcuding starttime) corresponds to the metadata of the
+            original traces if those are the same. Additionaly, the entries
+            ``stack`` (result of the format operation on the group parameter)
+            and ``stack_count`` (number of stacked traces)
+            are written to the stats object(s).
+
+        >>> from obspy import read
+        >>> st = read()
+        >>> stack = st.stack()
+        >>> print(stack)  # doctest: +ELLIPSIS
+        1 Trace(s) in Stream:
+        BW.RJOB.. | 2009-08-24T00:20:03.000000Z - ... | 100.0 Hz, 3000 samples
+        """
+        if group == 'id':
+            group = '{network}.{station}.{location}.{channel}'
+        groups = collections.defaultdict(list)
+        for tr in self:
+            groups[group.format(**tr.stats)].append(tr)
+        stacks = []
+        for groupid, traces in groups.items():
+            header = {k: v for k, v in traces[0].stats.items()
+                      if all(tr.stats.get(k) == v for tr in traces)}
+            header['stack'] = groupid
+            header['stack_count'] = len(traces)
```

Comment:
We don't really have a good concept for this type of meta-data for the stack so I also don't really know what's the best way here. But maybe move these two to a new `stack` group in the `stats` dictionary? Really just a suggestion though.

Reply:
So, you would prefer `stats.stack.id/group` and `stats.stack.count`? I don't know, I think I would prefer slightly the flat variant, but I am also OK with this one.

Reply:
Hmm, one could also add 'stats.stack.type'.

Reply:
Yea that might work. I slightly prefer to group all the `stack` meta parameters which I think would look a bit nicer.